### PR TITLE
Release bulk api

### PR DESCRIPTION
**Pull request by mohammed-fahad**

### CODE REVIEW

File: src/Client/Client.php [hunk 1]
```diff
@@ -96,6 +100,113 @@ public function createObject(string $sObject = null, array $data = null)
...
+        return $result->get();
+    }
+
+    /**
+     * @param string|null $jobId
+     * @return mixed
+     * @throws \Salesforce\Client\Exception\ResultException
+     * @throws ClientException
+     */
+    public function closeBulkJob(string $jobId = null)
```
Comment (ngodinhloc) [inline src/Client/Client.php]: should we rename these functions to be  without "bulk" ?

File: src/Client/Client.php [hunk 2]
```diff
@@ -28,6 +29,9 @@ class Client
     protected $logger;
 
     const MSG_DEBUG_CREATE_START = 'Start creating object in Salesforce. Type: %s. Data: %s';
+    const MSG_DEBUG_CREATE_BULK_START = 'Start creating object in Salesforce. Type: %s. Action: %s';
+    const MSG_DEBUG_ADD_BATCHES_TO_BULK_START = 'Start adding batches to bulk in Salesforce. JobId: %s. Data: %s';
+    const MSG_DEBUG_CLOSE_BULK_START = 'Start closing bulk Job in Salesforce. JobId: %s.';
```
Comment (ngodinhloc) [inline src/Client/Client.php]: should we rename these constants to be without "BULK" and "bulk" ?

File: src/Client/Client.php [hunk 3]
```diff
@@ -96,6 +100,113 @@ public function createObject(string $sObject = null, array $data = null)
...
+        $result = new Result($response);
+
+        return $result->get();
+    }
+
+    /**
+     * @param string|null $jobId
+     * @return mixed
+     * @throws \Salesforce\Client\Exception\ResultException
+     * @throws ClientException
```
Comment (ngodinhloc) [inline src/Client/Client.php]: please thrown/return full namespace

File: src/Job/Job.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    /** @var String */
+    protected $id;
+
+    /** @var String */
+    protected $operation;
+
+    /** @var String */
+    protected $object;
+
+    /** @var Entity */
```
Comment (ngodinhloc) [inline src/Job/Job.php]: full namespace

File: src/Job/Job.php [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * @return Connection
+     */
+    public function getConnection(): Connection
+    {
+        return $this->connection;
+    }
+
+    /**
+     * @param Connection $connection
```
Comment (ngodinhloc) [inline src/Job/Job.php]: full namesapce

File: src/Job/JobManager.php [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+ */
+class JobManager
+{
+    /** @var Connection */
+    protected $connection;
+
+    /** @var Mapper */
+    protected $mapper;
+
+    /** @var EventDispatcherInterface */
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: please use full namespace in php docs

File: src/Job/JobManager.php [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    /**
+     * @param string $className
+     * @param Job $job
+     * @return Job
+     * @throws JobException
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     * @throws \Salesforce\ORM\Exception\MapperException
+     */
+    public function createJob(string $className, Job $job)
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: createJob function should not take a Job as parameter, it should take in any necessary parameters and return a new Job

File: src/Job/JobManager.php [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        return $job;
+    }
+
+    /**
+     * @param Job $job
+     * @param array $header
+     * @param array $data
+     * @return Job
+     * @throws JobException
+     * @throws EntityException
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: full namespace

File: src/Job/JobManager.php [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    {
+        $this->validateJobData($job, $header, $data);
+
+        //load the CSV document from a string
+        $csv = Writer::createFromString('');
+        $csv->insertOne($header);
+        $csv->insertAll($data);
+
+        $jobAddedSuccessfully = $this->connection->getClient()->addToBulkJobBatches($job->getId(), $csv->getContent());
+
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: from line 104-109 : should be in a separate function name "batchJob()" or "batchJobData()".
processJob will then rely on $job->getType and/or $job->getAction to perform the suitable action.
For example:
switch($job->getType){
   case Constants::JOB_BULK_UPSERT:
         $result = $this->batchJob($job);
   break;
}

File: src/Job/JobManager.php [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    /**
+     * @param Job $job
+     * @return Job
+     * @throws JobException
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     */
+    public function closeJob(Job $job)
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: closeJob function should not return a Job, it should change status of $job:
public function closeJob(Job &$job){}

File: src/Job/JobManager.php [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        return $job;
+    }
+
+    /**
+     * @param Job $job
+     * @return mixed
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     */
+    public function getJobInfo(Job &$job)
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: no need to return Job as this function already take reference of job.

File: src/Job/JobManager.php [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+        return $jobInfo;
+    }
+
+    /**
+     * @param Job $job
+     * @return JobResult
+     */
+    public function getJobResult(Job $job)
+    {
+        return new JobResult($job, $this->connection);
```
Comment (ngodinhloc) [inline src/Job/JobManager.php]: this function should do more than just return new JobResult
All the stuffs dealing with $this->connection should happen inside JobManager, JobResult only take the returned results, which could be an response interface or array (similar Client/Result) and then interpret that result to return the expected format

File: src/Job/JobManager.php (hunk 11 above)
Comment (mohammed-fahad) [reply src/Job/JobManager.php]: it returns the job since it has more info that still not part of Job object.

File: src/Job/JobResult.php [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @return array
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     */
+    public function getSuccessFullResult()
+    {
+        $passedResult =$this->connection->getClient()->bulkJobGet(JobConstants::JOB_INGEST_ENDPOINT . $this->job->getId() . '/' . JobConstants::JOB_RESULT_PASSED_RESULT_ENDPOINT);
+
```
Comment (ngodinhloc) [inline src/Job/JobResult.php]: these stuffs should happen inside JobManager

File: src/Job/JobResult.php [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @return array
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     */
+    public function getFailedFullResult()
+    {
+        $failedResult = $this->connection->getClient()->bulkJobGet(JobConstants::JOB_INGEST_ENDPOINT . $this->job->getId() . '/' . JobConstants::JOB_RESULT_FAILED_RESULT_ENDPOINT);
+
```
Comment (ngodinhloc) [inline src/Job/JobResult.php]: these stuffs should happen inside JobManager

File: src/Job/JobResult.php [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @return array
+     * @throws \Salesforce\Client\Exception\ClientException
+     * @throws \Salesforce\Client\Exception\ResultException
+     */
+    public function getUnprocessedRecord()
+    {
+        $unprocessedRecords = $this->connection->getClient()->bulkJobGet(JobConstants::JOB_INGEST_ENDPOINT . $this->job->getId() . '/' . JobConstants::JOB_RESULT_UNPROCESSED_RESULT_ENDPOINT);
+
```
Comment (ngodinhloc) [inline src/Job/JobResult.php]: same here

File: src/Job/JobManager.php (hunk 7 above)
Comment (mohammed-fahad) [reply src/Job/JobManager.php]: I we keep adding parameters for things like External Id than the function parameters will keep growing. If we pass on a job than we have the flexibility to add additional fields easily! What you reckon ?

Comment (ngodinhloc) [reply src/Job/JobManager.php]: if that the case then the function name is not good. It should be "registerJob" or something

File: src/Job/JobManager.php (hunk 11 above)
Comment (ngodinhloc) [reply src/Job/JobManager.php]: still it only returns the job object so whatever info returned is in that job object, should be set on that object

**Comment by darkbluesun** [comment]
Hi @mohammed-fahad, it looks like you've got no tests to support this diff. Can you please write some? Thanks.

File: tests/Client/ClientTest.php [hunk 16]
```diff
@@ -146,6 +146,25 @@ public function testUpdateObject()
...
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::postJson()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::putCsv()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::setParamsFromAccessToken()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getOAuthAccessToken()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getAuthorizationHeader()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getClientCredentialsAccessToken()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getPasswordAccessToken()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getRefreshToken()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getOAuthAccessTokenFromResponse()
+     * @covers \Salesforce\Restforce\ExtendedOAuthRestClient::getNewToken()
```
Comment (darkbluesun) [inline tests/Client/ClientTest.php]: That's a lot of things for one test to cover. Is this a unit test or a functional test actually?

Comment (mohammed-fahad) [reply tests/Client/ClientTest.php]: The work is still in progress! These all not suppose to be here.

**Pull request merged**